opm/grid/cpgrid/LgrOutputHelpers.hpp

```diff
@@ -0,0 +1,169 @@
+void extractSolutionLevelGrids(const Dune::CpGrid& grid,
+                               const Opm::data::Solution& leafSolution,
+                               std::vector<Opm::data::Solution>&);
+
+/// @brief Constructs restart-value containers for all grid refinement levels.
+///
+/// The level-specific solution data are first derived from the leaf solution
+/// using extractSolutionLevelGrids(...). Other data components (such as wells,
+/// group/network values, and aquifers) are passed unchanged to each level.
+///
+/// @param [template] Grid The function has no effect for grids other than CpGrid.
+/// @param [in]       grid
+/// @param [in]       leafRestartValue
+/// @param [out]      A vector of RestartValue objects, one for each refinement level
+///                   (from level 0 to grid.maxLevel()).
+template <typename Grid>
+void extractRestartValueLevelGrids(const Grid& grid,
+                                   const Opm::RestartValue& leafRestartValue,
+                                   std::vector<Opm::RestartValue>& restartValue_levels);
+
+} // namespace Lgr
+} // namespace Opm
+
+template <typename Container>
+Container Opm::Lgr::reorderForOutput(const Container& simulatorContainer,
+                                     const std::vector<int>& toOutput)
+{
+    // Use toOutput to reorder simulatorContainer
+    Container outputContainer;
+    outputContainer.resize(toOutput.size());
+    for (std::size_t i = 0; i < toOutput.size(); ++i) {
+        outputContainer[i] = simulatorContainer[toOutput[i]];
+    }
+    return outputContainer;
+}
+
+template <typename Grid>
+void Opm::Lgr::extractRestartValueLevelGrids(const Grid& grid,
+                                             const Opm::RestartValue& leafRestartValue,
+                                             std::vector<Opm::RestartValue>& restartValue_levels)
+{
+    if constexpr (std::is_same_v<Grid, Dune::CpGrid>) {
+
+        int maxLevel = grid.maxLevel();
+        restartValue_levels.resize(maxLevel+1); // level 0, 1, ..., max level
+
+        std::vector<Opm::data::Solution> dataSolutionLevels{};
+        extractSolutionLevelGrids(grid,
+                                  leafRestartValue.solution,
+                                  dataSolutionLevels);
+
+        for (int level = 0; level <= maxLevel; ++level) {
+            restartValue_levels[level] = Opm::RestartValue(dataSolutionLevels[level],
+                                                           leafRestartValue.wells,
+                                                           leafRestartValue.grp_nwrk,
+                                                           leafRestartValue.aquifer,
+                                                           level);
+        }
+
+        for (const auto& [rst_key, leafVector] : leafRestartValue.extra) {
+
+            std::vector<std::vector<double>> levelVectors{};
+            levelVectors.resize(maxLevel+1);
+
+            const auto rubbish = -1;
```

Comment:
This is only used once and this variable should be omitted.

Should we rather use a different value (numeric_limits<double>::max())?